run/add_ddos_protection_iptables_rule.sh

```diff
@@ -204,64 +147,61 @@ if [ ! -z ${TEST_MODE} ]; then
-    if [ ! -z ${CONN_LOGGING_LEVEL} ]; then
-        RULE="${LOG_CHAIN} -j LOG --log-level ${CONN_LOGGING_LEVEL} --log-prefix \"connlimit: \""
-        ${OPERATION} "${RULE}"
-    fi
+# Make sure the previous default logging rule is removed.  It causes too much CPU overhead under load.
+RULE="${LOG_CHAIN} -j LOG --log-level warning --log-prefix \"connlimit: \""
+delete_rule ${RULE}
 
-    # Append a rule that finally rejects connection
-    RULE="${LOG_CHAIN} -p tcp -j REJECT --reject-with tcp-reset"
-    ${OPERATION} "${RULE}"
-
-    # Append a rule to limit the total number of simultaneous client connections
-    RULE="${IP_TABLES_CHAIN} -p tcp --syn --dport ${DPORT} -m connlimit --connlimit-above ${OVER_ALL_CONN_LIMIT} --connlimit-mask 0 -j ${LOG_CHAIN}"
-    ${OPERATION} "${RULE}"
-
-    # Append a rule to limit the number connections per IP address
-    RULE="${IP_TABLES_CHAIN} -p tcp -m tcp --dport ${DPORT} --tcp-flags FIN,SYN,RST,ACK SYN -m connlimit --connlimit-above ${CONN_LIMIT_PER_IP} --connlimit-mask 32 --connlimit-saddr -j ${LOG_CHAIN}"
-    ${OPERATION} "${RULE}"
-
-    # Append rules to rate limit connections
-    if (( ${CONN_RATE_LIMIT_LIMIT} > 0 && ${CONN_RATE_LIMIT_PERIOD} > 0 )); then
-        echo "Including settings for rate limiting ..."
-        RULE="${IP_TABLES_CHAIN} -p tcp -m tcp --dport ${DPORT} -m conntrack --ctstate NEW -m recent --set --name DEFAULT --mask 255.255.255.255 --rsource"
-        ${OPERATION} "${RULE}"
-        RULE="${IP_TABLES_CHAIN} -p tcp -m tcp --dport ${DPORT} -m conntrack --ctstate NEW -m recent --update --seconds ${CONN_RATE_LIMIT_PERIOD} --hitcount ${CONN_RATE_LIMIT_LIMIT} --name DEFAULT --mask 255.255.255.255 --rsource -j ${LOG_CHAIN}"
-        ${OPERATION} "${RULE}"
-    else
-        echo "Rate limiting is disabled, skipping settings for rate limiting ..."
-    fi
+# Append a rule that sets log level and log prefix
+# Default to no logging unless a logging level is explicitly supplied.
+if [ ! -z ${CONN_LOGGING_LEVEL} ]; then
+    RULE="${LOG_CHAIN} -j LOG --log-level ${CONN_LOGGING_LEVEL} --log-prefix \"connlimit: \""
+    ${OPERATION} ${RULE}
+fi
 
-    if delete; then
-        # Remove logging chain for rejected connections
-        iptables -X ${LOG_CHAIN} 2>/dev/null 1>&2
-    fi
+# Append a rule that finally rejects connection
+RULE="${LOG_CHAIN} -p tcp -j REJECT --reject-with tcp-reset"
+make_last_rule ${RULE}
+
+# Append a rule to limit the total number of simultaneous client connections
+RULE="${IP_TABLES_CHAIN} -p tcp --syn --dport ${DPORT} -m connlimit --connlimit-above ${OVER_ALL_CONN_LIMIT} --connlimit-mask 0 -j ${LOG_CHAIN}"
+${OPERATION} ${RULE}
+
+# Append a rule to limit the number connections per IP address
+RULE="${IP_TABLES_CHAIN} -p tcp -m tcp --dport ${DPORT} --tcp-flags FIN,SYN,RST,ACK SYN -m connlimit --connlimit-above ${CONN_LIMIT_PER_IP} --connlimit-mask 32 --connlimit-saddr -j ${LOG_CHAIN}"
```

Comment:
With `--connlimit-mask 32` we bind only to a single ip adress like `10.0.0.1/32` probably we can think about whitelisting ranges with a `--connlimit-mask 8` would result in `10.0.0.0/8`

Reply:
As far as I understand this script (by @WadeBarnes ): the connlimit-mask sets the connection limit per this range of IPs and we do want to limit the connections per single IP, not per IP range.
AFAIK this is what the VM operators do right now -> let's do it the same way or discuss with everybody and chage for all